File: app/src/main/java/top/witzzz/moonbazaar/data/MoonWorkerApi.kt
package top.witzzz.moonbazaar.data

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * 对 MoonBazaar Worker 的轻量 HTTP 客户端（session 鉴权版）。
 *
 *  - App 一律请求 [AppConfig.BASE_URL]（自己部署的 Cloudflare Worker），由 Worker 转发
 *    到 MoonBazaar 上游并自动注入 X-API-Key / Authorization(Bearer)。
 *  - 客户端身份使用 Worker 发放的 session token（见 [Session]），每次请求自动带：
 *        X-Client-Session: <session_token>
 *  - 写操作统一走 [proxyPost]，并强制携带 [idempotencyKey]（MoonBazaar 要求 8-128 字符）。
 */
class MoonWorkerApi {

    private val jsonType = "application/json; charset=utf-8".toMediaType()

    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(45, TimeUnit.SECONDS)
        .writeTimeout(45, TimeUnit.SECONDS)
        .build()

    data class CallResult(
        val ok: Boolean,               // HTTP 是否 2xx
        val httpStatus: Int = -1,
        val body: String = "",         // Worker / 上游原始响应文本
        val replayed: Boolean = false, // Idempotent-Replayed == true
        val thrown: String? = null     // 本地网络/IO 错误
    ) {
        val json: JSONObject?
            get() = try {
                if (body.isBlank()) null else JSONObject(body)
            } catch (_: Exception) {
                null
            }

        val upError: JSONObject?
            get() = json?.optJSONObject("error")

        /** 对 UI 友好：优先取上游 error.message，其次本地 thrown/HTTP。 */
        val readable: String
            get() = upError?.optString("message")?.takeIf { it.isNotBlank() }
                ?: thrown
                ?: if (!ok) "HTTP $httpStatus" else ""
    }

    /** Worker 自带的只读接口（如 /status、/user/tokens）。 */
    fun get(path: String): CallResult = exec("GET", path, null, null)

    /** Worker 自带写接口（如 DELETE /user/tokens 登出）。 */
    fun del(path: String): CallResult = exec("DELETE", path, null, null)

    /** 经 Worker 代理的只读 GET：path 不带 /proxy 前缀，如 account/status。 */
    fun proxyGet(path: String): CallResult = exec("GET", "/proxy/$path", null, null)

    /** 经 Worker 代理的写操作，body 可为 null，需幂等键。path 不带 /proxy。 */
    fun proxyPost(path: String, body: JSONObject?, idempotencyKey: String): CallResult =
        exec("POST", "/proxy/$path", body, idempotencyKey)

    /** 经 Worker 代理、**无需幂等键**的写操作（如 account/email-verification）。 */
    fun proxyPostNoIdem(path: String, body: JSONObject?): CallResult =
        exec("POST", "/proxy/$path", body, null)

    private fun exec(method: String, urlPath: String, body: JSONObject?, idem: String?): CallResult {
        return try {
            val url = AppConfig.BASE_URL.trimEnd('/') + urlPath
            val b = Request.Builder()
                .url(url)
                .header("X-Source-App", AppConfig.SOURCE_APP)
                .header("Accept", "application/json")
            val sess = Session.tokenValue
            if (sess != null) b.header("X-Client-Session", sess)
            if (idem != null) b.header("Idempotency-Key", idem)

            val req = when (method.uppercase()) {
                "GET" -> b.get().build()
                "DELETE" -> b.delete().build()
                else -> {
                    val rb = if (body == null) "{}".toRequestBody(jsonType)
                    else body.toString().toRequestBody(jsonType)
                    b.post(rb).build()
                }
            }

            client.newCall(req).execute().use { resp ->
                CallResult(
                    ok = resp.isSuccessful,
                    httpStatus = resp.code,
                    body = resp.body?.string() ?: "",
                    replayed = resp.header("Idempotent-Replayed") == "true"
                )
            }
        } catch (e: Exception) {
            CallResult(ok = false, httpStatus = -1, body = "",
                thrown = "网络请求失败：${e.message}")
        }
    }
}

File: ios/MoonBazaar/MoonWorkerApi.swift
import Foundation

/// 对 MoonBazaar Worker 的轻量 HTTP 客户端（session 鉴权版）。
///
/// - App 一律请求 `AppConfig.BASE_URL`（自己部署的 Cloudflare Worker），由 Worker 转发到
///   MoonBazaar 上游并自动注入 `X-API-Key` / `Authorization(Bearer)`。
/// - 客户端身份使用 Worker 发放的 session token（见 `Session`），每次请求自动带
///   `X-Client-Session: <session_token>`。
/// - 写操作统一走 `proxyPost`，并强制携带 `idempotencyKey`（MoonBazaar 要求 8–128 字符）。
///
/// 说明：`await session.data(for:)` 是挂起而不是阻塞，因此标 `@MainActor` 不会卡住 UI，
/// 同时保证 `Session` 的 `@Published` 更新始终在主线程。
@MainActor
final class MoonWorkerApi {

    private let session: URLSession

    /// 与 Android OkHttp 的 connect/read/write 超时保持一致。
    init() {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 45
        cfg.timeoutIntervalForResource = 45
        cfg.requestCachePolicy = .reloadIgnoringLocalCacheData
        cfg.httpAdditionalHeaders = ["Accept": "application/json"]
        session = URLSession(configuration: cfg)
    }

    /// 一次调用的结果。
    struct CallResult {
        var ok: Bool                  // HTTP 是否 2xx
        var httpStatus: Int = -1
        var body: String = ""         // Worker / 上游原始响应文本
        var replayed: Bool = false    // Idempotent-Replayed == true
        var thrown: String?           // 本地网络 / IO 错误

        var json: JObj? {
            guard !body.isEmpty, let data = body.data(using: .utf8) else { return nil }
            return (try? JSONSerialization.jsonObject(with: data)) as? JObj
        }

        var upError: JObj? {
            guard let j = json, let e = j["error"] else { return nil }
            return J.dict(e)
        }

        /// 对 UI 友好：优先取上游 error.message，其次本地 thrown / HTTP。
        var readable: String {
            if let e = upError, let m = e["message"] as? String, !m.isEmpty { return m }
            if let t = thrown, !t.isEmpty { return t }
            if !ok { return "HTTP \(httpStatus)" }
            return ""
        }
    }

    // MARK: - 便捷入口

    /// Worker 自带的只读接口（如 `/status`）。
    func get(_ path: String) async -> CallResult { await exec("GET", path, nil, nil) }

    /// Worker 自带写接口（如 `DELETE /user/tokens` 登出）。
    func del(_ path: String) async -> CallResult { await exec("DELETE", path, nil, nil) }

    /// 经 Worker 代理的只读 GET：path 不带 `/proxy` 前缀，如 `account/status`。
    func proxyGet(_ path: String) async -> CallResult { await exec("GET", "/proxy/\(path)", nil, nil) }

    /// 经 Worker 代理的写操作，body 可为 nil，需幂等键。path 不带 `/proxy`。
    func proxyPost(_ path: String, _ body: JObj?, idempotencyKey: String) async -> CallResult {
        await exec("POST", "/proxy/\(path)", body, idempotencyKey)
    }

    /// 经 Worker 代理、**无需幂等键**的写操作（如 `account/email-verification`）。
    func proxyPostNoIdem(_ path: String, _ body: JObj?) async -> CallResult {
        await exec("POST", "/proxy/\(path)", body, nil)
    }

    // MARK: - 核心

    private func exec(_ method: String, _ urlPath: String, _ body: JObj?, _ idem: String?) async -> CallResult {
        let base = AppConfig.BASE_URL.hasSuffix("/")
            ? String(AppConfig.BASE_URL.dropLast())
            : AppConfig.BASE_URL
        guard let url = URL(string: base + urlPath) else {
            return CallResult(ok: false, thrown: "无效的请求地址：\(base + urlPath)")
        }

        var req = URLRequest(url: url)
        req.httpMethod = method.uppercased()
        req.setValue(AppConfig.SOURCE_APP, forHTTPHeaderField: "X-Source-App")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if let sess = Session.shared.tokenValue, !sess.isEmpty {
            req.setValue(sess, forHTTPHeaderField: "X-Client-Session")
        }
        if let idem = idem {
            req.setValue(idem, forHTTPHeaderField: "Idempotency-Key")
        }

        if method.uppercased() == "POST" {
            let payload: Data
            if let body = body, let data = try? JSONSerialization.data(withJSONObject: body) {
                payload = data
            } else {
                payload = Data("{}".utf8)
            }
            req.httpBody = payload
            req.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        }

        do {
            let (data, response) = try await session.data(for: req)
            let http = response as? HTTPURLResponse
            let code = http?.statusCode ?? -1
            return CallResult(
                ok: (200...299).contains(code),
                httpStatus: code,
                body: String(data: data, encoding: .utf8) ?? "",
                replayed: http?.value(forHTTPHeaderField: "Idempotent-Replayed") == "true",
                thrown: nil
            )
        } catch {
            return CallResult(ok: false, httpStatus: -1, body: "",
                              thrown: "网络请求失败：\(error.localizedDescription)")
        }
    }
}